execution/aggregate/khashaggregate.go

```diff
@@ -82,25 +82,28 @@ func (a *kAggregate) Next(ctx context.Context) ([]model.StepVector, error) {
 	if err != nil {
 		return nil, err
 	}
-	if in == nil {
-		return nil, nil
-	}
-
-	defer a.next.GetPool().PutVectors(in)
-
 	args, err := a.paramOp.Next(ctx)
 	if err != nil {
 		return nil, err
 	}
-	for i := range a.params {
-		a.params[i] = math.NaN()
-		if i < len(args) {
-			a.params[i] = args[i].Samples[0]
-			a.paramOp.GetPool().PutStepVector(args[i])
+
+	for i := range args {
+		a.params[i] = args[i].Samples[0]
+		a.paramOp.GetPool().PutStepVector(args[i])
+
+		val := a.params[i]
+		if val > math.MaxInt64 || val < math.MinInt64 || math.IsNaN(val) {
+			return nil, errors.Newf("scalar value %v overflows int64", val)
+		}
+		if int(val) == 0 {
+			return nil, nil
 		}
 	}
 	a.paramOp.GetPool().PutVectors(args)
 
+	if in == nil {
```

Comment:
Why do we need to move this check here?

Reply:
That is needed for this test to pass: 
```
		{
			name:  "topk with NaN and no matching series",
			query: "topk(NaN, not_there)",
			start: time.Unix(0, 0),
			end:   time.Unix(3000, 0),
			step:  2 * time.Second,
		},

```
upstream prometheus errors out even if there is no matching series to compute topk for, if we return early we dont get that error.

Reply:
Got it 👍